### PR TITLE
Assist: add posthog event

### PR DESCRIPTION
**Pull request by justinas**
Adds reporting of "Assist completion" events to PostHog.

"Completion tokens" are tallied up by counting the items in the completion stream. To estimate "prompt tokens", I included the Go port of the `tiktoken` library (OpenAI does not send a `usage` field in stream responses).

Corresponding Cloud changes: https://github.com/gravitational/cloud/pull/4392

### CODE REVIEW

File: proto/prehog/v1alpha/teleport.proto [hunk 1]
```diff
@@ -361,6 +361,24 @@ message UICreateNewRoleViewDocumentationClickEvent {
   string user_name = 1;
 }
 
+// UICallToActionClickEvent is a click in a Teleport Web UI's CTA
```
Comment (justinas) [inline proto/prehog/v1alpha/teleport.proto:364]: These are unrelated changes: the types were newly introduced to Cloud in https://github.com/gravitational/cloud/pull/4326 , and will merged to Teleport master in https://github.com/gravitational/teleport/pull/25325 . Once that happens, we should merge master into `dev-ai` and these changes will "disappear".

File: proto/prehog/v1alpha/teleport.proto [hunk 2]
```diff
@@ -448,6 +475,10 @@ message SubmitEventRequest {
 
     UIDiscoverIntegrationAWSOIDCConnectEvent ui_discover_integration_aws_oidc_connect_event = 40;
     UIDiscoverDatabaseRDSEnrollEvent ui_discover_database_rds_enroll_event = 41;
+
+    UICallToActionClickEvent ui_call_to_action_click_event = 42;
```
Comment (justinas) [inline proto/prehog/v1alpha/teleport.proto:479]: Ditto

File: lib/web/assistant.go [hunk 3]
```diff
@@ -189,6 +191,12 @@ type partialFinalizePayload struct {
 
 // runAssistant upgrades the HTTP connection to a websocket and starts a chat loop.
 func runAssistant(h *Handler, w http.ResponseWriter, r *http.Request, sctx *SessionContext) error {
+	// Initialize a tokenizer for prompt token accounting
+	tokenizerInstance, err := tokenizer.Get(tokenizer.Cl100kBase)
```
Comment (jakule) [inline lib/web/assistant.go:195]: nit: Can you please add to the comment that Cl100k tokenizer is being used by GPT-3 and GPT-4? Without context, it's not clear why we're using this particular one.

File: lib/web/assistant.go [hunk 4]
```diff
@@ -275,6 +283,9 @@ func runAssistant(h *Handler, w http.ResponseWriter, r *http.Request, sctx *Sess
 		// write user message to both in-memory chain and persistent storage
 		chat.Insert(openai.ChatMessageRoleUser, wsIncoming.Payload)
 
+		// TODO(justinas): should we handle this error?
```
Comment (jakule) [inline lib/web/assistant.go:286]: Print it in logs (not as an error) and make sure that `promptTokens` is 0. I don't think that having (if this ever happens) a 99% accurate token usage is an issue.

File: lib/web/assistant.go [hunk 5]
```diff
@@ -284,9 +295,25 @@ func runAssistant(h *Handler, w http.ResponseWriter, r *http.Request, sctx *Sess
...
+					AssistCompletion: &usageeventsv1.AssistCompletionEvent{
+						TotalTokens:      int64(len(promptTokens) + completionTokens),
+						PromptTokens:     int64(len(promptTokens)),
+						CompletionTokens: int64(completionTokens),
+					},
+				},
+			},
+		}
+		if err := authClient.SubmitUsageEvent(r.Context(), usageEventReq); err != nil {
+			h.log.WithError(err).Error("Failed to emit usage event")
```
Comment (jakule) [inline lib/web/assistant.go:315]: Can you please change the log level to waning or debug? This is not a critical error, and this will print ERROR message in our logs and the whole call stack. It's hard to debug errors if non-critical logs are marked as errors. Our users are also very often confused by that.

File: proto/prehog/v1alpha/teleport.proto (hunk 1 above)
Comment (jakule) [reply proto/prehog/v1alpha/teleport.proto:364]: @justinas I updated the `dev-ai` with the latest master. Unfornetly, GH is not able to resolve the conflicts automatically, as all PRs are merged with squash which changes the commits IDs.

File: lib/ai/chat.go [hunk 6]
```diff
@@ -136,16 +138,17 @@ func (chat *Chat) Complete(ctx context.Context) (any, error) {
...
 
 	// fetch the first delta to check for a possible JSON payload
 	var response openai.ChatCompletionStreamResponse
 top:
 	response, err = stream.Recv()
 	if err != nil {
-		return nil, trace.Wrap(err)
+		return nil, numTokens, trace.Wrap(err)
 	}
+	numTokens++
```
Comment (jakule) [inline lib/ai/chat.go:151]: Why are we adding one to `numTokens`?

File: lib/ai/chat.go [hunk 7]
```diff
@@ -106,14 +106,16 @@ type StreamingMessage struct {
 }
 
 // Complete completes the conversation with a message from the assistant based on the current context.
-func (chat *Chat) Complete(ctx context.Context) (any, error) {
+func (chat *Chat) Complete(ctx context.Context) (any, int, error) {
```
Comment (jakule) [inline lib/ai/chat.go:109]: Can you please update the comment?

File: lib/web/assistant.go [hunk 8]
```diff
@@ -275,6 +283,9 @@ func runAssistant(h *Handler, w http.ResponseWriter, r *http.Request, sctx *Sess
 		// write user message to both in-memory chain and persistent storage
 		chat.Insert(openai.ChatMessageRoleUser, wsIncoming.Payload)
 
+		// TODO(justinas): should we handle this error?
+		promptTokens, _, _ := tokenizerInstance.Encode(wsIncoming.Payload)
```
Comment (jakule) [inline lib/web/assistant.go:287]: This is not accurate.  `Payload` contains only the string from a user. At the beginning of each conversation, we add out system prompt that is not included here https://github.com/gravitational/teleport/blob/9c405c7fd59bd1144a207538b3148e44d4208167/lib/ai/prompt.go#L26.
Then each query contains all previous messages and responses (to remember the context).

File: api/proto/teleport/usageevents/v1/usageevents.proto [hunk 9]
```diff
@@ -285,6 +285,16 @@ message UICreateNewRoleCancelClickEvent {}
 // UICreateNewRoleViewDocumentationClickEvent is an event that can be triggered during custom role creation
 message UICreateNewRoleViewDocumentationClickEvent {}
 
+// AssistCompletionEvent is an event that is emitted when a completion occurs in the Assistant
+message AssistCompletionEvent {
+  // TotalTokens is the total amount of token used to satisfy this request
+  int64 total_tokens = 1;
```
Comment (jakule) [inline api/proto/teleport/usageevents/v1/usageevents.proto:291]: Any reasons not to use `unit64`? The number of tokens will never be negative.

Comment (espadolini) [reply api/proto/teleport/usageevents/v1/usageevents.proto:291]: +1 (although in terms of wire format it's the same thing)

File: lib/web/assistant.go (hunk 5 above)
Comment (espadolini) [reply lib/web/assistant.go:315]: +1 for warning, we don't even log at error level if we fail to emit audit log events, most of the time.

File: lib/ai/chat.go (hunk 6 above)
Comment (justinas) [reply lib/ai/chat.go:151]: I was under the impression that every item on the stream (except for `[DONE]`) is a single token. Is the start of the stream also special?

File: lib/web/assistant.go (hunk 8 above)
Comment (justinas) [reply lib/web/assistant.go:315]: Updated the calculation to take history into the account, made other adjustments to match the OpenAI calculations.

File: api/proto/teleport/usageevents/v1/usageevents.proto (hunk 9 above)
Comment (justinas) [reply api/proto/teleport/usageevents/v1/usageevents.proto]: 🤷 Doesn't look like we use unsigned ints anywhere in these protos, even when the value is obviously non-negative, e.g. `resource_count`.

File: proto/prehog/v1alpha/teleport.proto [hunk 10]
```diff
@@ -405,6 +405,20 @@ message AgentMetadataEvent {
...
+// i.e. a user submits a prompt, and receives an answer from the Assist
+message AssistCompletionEvent {
+  // anonymized
+  string user_name = 1;
+
+  string conversation_id = 2;
+
+  int64 total_tokens = 3;
+  int64 prompt_tokens = 4;
+  int64 completion_tokens = 5;
```
Comment (jakule) [inline proto/prehog/v1alpha/teleport.proto:419]: Can you add comments here?

**Review by jakule — APPROVED** [review]
Looks good.

**Comment by jakule** [comment]
@justinas I fixed the failing linter on `dev-ai` after you rebase this branch CI should be happy.

File: proto/prehog/v1alpha/teleport.proto (hunk 10 above)
Comment (espadolini) [reply proto/prehog/v1alpha/teleport.proto]: Remember to sync the changes with gravitational/cloud!

File: lib/ai/chat.go (hunk 6 above)
Comment (xacrimon) [reply lib/ai/chat.go:180]: I believe you can get an empty delta sometimes, why I am still not sure of. Could you add a check or verify it's always one?

**Review by xacrimon — APPROVED** [review]

**Comment by justinas** [comment]
> I believe you can get an empty delta sometimes, why I am still not sure of. Could you add a check or verify it's always one?

@xacrimon Did some manual API calls, always got a delta with `{"role": "assistant"}` first. AFAIK that would still be considered a token, so the current logic should be fine?
